src/runtimes/runners/runWithDefaults.ts

```diff
@@ -0,0 +1,122 @@
+/*!--------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { AccumulatorStream, ClientIdentity, GeneratorCommandResponse, IContainersClient, isChildProcessError, Like, normalizeCommandResponseLike, PromiseCommandResponse, ShellStreamCommandRunnerFactory, ShellStreamCommandRunnerOptions, VoidCommandResponse } from '../docker';
+import { ext } from '../../extensionVariables';
+import { RuntimeManager } from '../RuntimeManager';
+import { withDockerEnvSettings } from '../../utils/withDockerEnvSettings';
+
+type ClientCallback<TClient, T> = (client: TClient) => Like<PromiseCommandResponse<T>>;
+type VoidClientCallback<TClient> = (client: TClient) => Like<VoidCommandResponse>;
+type StreamingClientCallback<TClient, T> = (client: TClient) => Like<GeneratorCommandResponse<T>>;
+
+// 'env', 'shell', 'shellProvider', 'stdErrPipe', and 'strict' are set by this function and thus should not be included as arguments to the additional options
+type DefaultEnvStreamCommandRunnerOptions = Omit<ShellStreamCommandRunnerOptions, 'env' | 'shell' | 'shellProvider' | 'stdErrPipe' | 'strict'>;
+
+export async function runWithDefaultsInternal<T>(callback: ClientCallback<IContainersClient, T>, additionalOptions?: DefaultEnvStreamCommandRunnerOptions): Promise<T>;
+export async function runWithDefaultsInternal(callback: VoidClientCallback<IContainersClient>, additionalOptions?: DefaultEnvStreamCommandRunnerOptions): Promise<void>;
+export async function runWithDefaultsInternal<T>(callback: ClientCallback<IContainersClient, T> | VoidClientCallback<IContainersClient>, additionalOptions?: DefaultEnvStreamCommandRunnerOptions): Promise<T | void> {
+    return await runWithDefaults(
+        callback,
+        ext.runtimeManager,
+        additionalOptions
+    );
+}
+
+export function streamWithDefaultsInternal<T>(callback: StreamingClientCallback<IContainersClient, T>, additionalOptions?: DefaultEnvStreamCommandRunnerOptions): AsyncGenerator<T> {
+    return streamWithDefaults(
+        callback,
+        ext.runtimeManager,
+        additionalOptions
+    );
+}
+
+async function runWithDefaults<TClient extends ClientIdentity, T>(

```

Comment:
It might be worth defining the method overloads so the type system can properly determine the expected return type based on the value of callback:
```
function runWithDefaults<TClient extends ClientIdentity, T>(callback: ClientCallback<TClient, T>, runtimeManager: RuntimeManager<TClient>, additionalOptions?: DefaultEnvStreamCommandRunnerOptions): Promise<T>;
function runWithDefaults<TClient extends ClientIdentity>(callback: VoidClientCallback<TClient>, runtimeManager: RuntimeManager<TClient>, additionalOptions?: DefaultEnvStreamCommandRunnerOptions): Promise<void>;
async function runWithDefaults<TClient extends ClientIdentity, T>(callback: ClientCallback<TClient, T> | VoidClientCallback<TClient>, runtimeManager: RuntimeManager<TClient>, additionalOptions?: DefaultEnvStreamCommandRunnerOptions): Promise<T | void> { ...
```

Reply:
The definitions of the `runWithDefaultsInternal` / `streamWithDefaultsInternal` accomplish that

Reply:
So they do; should the names of runWithDefaults and runWithDefaultsInternal be swapped? I feel like the current naming is a bit confusing.

Reply:
I think that's reasonable.